### ISSUE TITLE
Bug: Crowdsale Tx's being classified as simple sends via rpc

### ISSUE THREAD

**Issue by achamely** [OmniLayer/omnicore#235]
@dexX7 @zathras-crypto 

example tx :  9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d
omnicored commit: 51f11fb reports it as a simple send where mastercored 0.0.9.3 reports it as a crowdsale

Output from mastercore 0.0.9.3

gettransaction_MP 9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d
{
  "error": null,
  "id": null,
  "result": {
    "amount": "2.00000000",
    "blocktime": 1396067389,
    "confirmations": 81038,
    "divisible": true,
    "fee": 0.0002,
    "ismine": false,
    "issuertokens": "160",
    "propertyid": 2,
    "purchasedpropertydivisible": false,
    "purchasedpropertyid": 2147483652,
    "purchasedpropertyname": "MCHEST Test Property",
    "purchasedtokens": "3214",
    "referenceaddress": "1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb",
    "sendingaddress": "1EqTta1Rt8ixAA32DuC29oukbsSWU62qAV",
    "txid": "9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d",
    "type": "Crowdsale Purchase",
    "type_int": 0,
    "valid": true,
    "version": 0
  }
}

Output from omnicored commit: 51f11fb
gettransaction_MP 9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d
{
  "error": null,
  "id": null,
  "result": {
    "amount": "2.00000000",
    "block": 293017,
    "blockhash": "0000000000000000c7f1a7fcc5de5bcf90b9ad9774fab78ed925cc60e45b51f3",
    "blocktime": 1396067389,
    "confirmations": 81038,
    "divisible": true,
    "fee": "0.00020000",
    "ismine": false,
    "propertyid": 2,
    "referenceaddress": "1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb",
    "sendingaddress": "1EqTta1Rt8ixAA32DuC29oukbsSWU62qAV",
    "txid": "9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d",
    "type": "Simple Send",
    "type_int": 0,
    "valid": true,
    "version": 0
  }
}

**Comment by achamely** [comment]
i'm currently updating/pulling from tip, recompiling to see if it's cleared up in any recent updates.

**Comment by achamely** [comment]
build from tip also exhibits same issue

**Comment by dexX7** [comment]
Do you know by any chance, if this is solely an issue with the, or do you think it's going deeper, and it was not counted as participation at all? I'm not sure what's the best way to determine it without headaches, but say for example you know it, because the balance is correct/incorrect.

I'm looking into it.

**Comment by dexX7** [comment]
Is this issue somehow limited to specific properties (e.g. Test Omni), or applies to every crowdsale participation?

`"omni_getactivecrowdsales"` could be used  to show currently active crowdsales, and `"omni_getcrowdsale" <propertyid> true` to list crowdsale participations. You could then test some results with `"omni_gettransaction <txid>"`.

You may also try `"omni_getcrowdsale" 2147483652 true` and then check, whether the transaction [9f8f19..6d](http://omnichest.info/lookuptx.aspx?txid=9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d) is shown.

**Comment by achamely** [comment]
yes, the 9f8f19 does show up in the list when calling "omni_getcrowdsale" 2147483652 true

did some spot checking,  a few of the maidsafe crowdsale purchases show up correctly and report correctly.   checked one or two other 'test' properties and sure enough it seems (at least right now)  that its limited to test property transactions ("246a3d80f56599ff192d6845b440651233a0727df2d32f8eb11d3ab5002b86f4")


It also appears as if internal balances are displayed correctly (getallbalancesforaddress_MP ) between 0.0.9.3 and 0.0.10 tip for addresses in those tx's

so issues is specific to the rpc return/classification of the tx

**Comment by dexX7** [comment]
> yes, the 9f8f19 does show up in the list when calling "omni_getcrowdsale" 2147483652 true

Hmm.. at first I thought this might be caused by some consensus rule changes, which affect the test ecosystem, because the MetaDEx or "send all" transactions are enabled for test properties, but not yet allowed with main ecosystem properties. Master Core might have processed these test transactions differently, because the logic didn't exist back then.

But since the transaction also shows up as actual participation (via `omni_getcrowdsale`), it means Omni Core considers the participation as valid.

At the very moment I'm without fully synced mainnet chain, so I'm unfortunally unable to check the actual transaction.

**Comment by dexX7** [comment]
I have a somewhat up-to-date version of Omni Core on a server (can't locate commit though), and I can confirm the issue:
```
bdaemon@master:~$ omnicore-cli omni_gettransaction 9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d
{
    "txid" : "9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d",
    "sendingaddress" : "1EqTta1Rt8ixAA32DuC29oukbsSWU62qAV",
    "referenceaddress" : "1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb",
    "ismine" : false,
    "confirmations" : 81062,
    "fee" : "0.00020000",
    "blocktime" : 1396067389,
    "valid" : true,
    "version" : 0,
    "type_int" : 0,
    "type" : "Simple Send",
    "propertyid" : 2,
    "divisible" : true,
    "amount" : "2.00000000"
}
```
```
bdaemon@master:~$ omnicore-cli omni_getcrowdsale 2147483652 true
{
    "propertyid" : 2147483652,
    "name" : "MCHEST Test Property",
    "active" : false,
    "issuer" : "1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb",
    "propertyiddesired" : 2,
    "tokensperunit" : "1500",
    "earlybonus" : 10,
    "percenttoissuer" : 5,
    "starttime" : 1395973991,
    "deadline" : 1396500000,
    "amountraised" : "0.00000000",
    "tokensissued" : "3374",
    "addedissuertokens" : "0",
    "closedearly" : false,
    "maxtokens" : false,
    "participanttransactions" : [
        {
            "txid" : "9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d",
            "amountsent" : "2.00000000",
            "participanttokens" : "3214",
            "issuertokens" : "160"
        }
    ]
}
```

The tokens were credited as expected:
```
bdaemon@master:~$ omnicore-cli omni_getbalance 1EqTta1Rt8ixAA32DuC29oukbsSWU62qAV 2147483652
{
    "balance" : "3214",
    "reserved" : "0"
}
```

And the participation was also logged during the initial synchronization:
```
2015-08-10 13:14:23 ____________________________________________________________________________________________________________________________________
2015-08-10 13:14:23 parseTransaction(block=293017, 2014-03-29 04:29:49 idx= 88); txid: 9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d
2015-08-10 13:14:23 	------------------------------
2015-08-10 13:14:23 	         version: 0, class B
2015-08-10 13:14:23 	            type: 0 (Simple Send)
2015-08-10 13:14:23 	        property: 2 (TMSC)
2015-08-10 13:14:23 	           value: 2.00000000
2015-08-10 13:14:23 update_tally_map(1EqTta1Rt8ixAA32DuC29oukbsSWU62qAV, 2=0x2, -200000000, ttype=0): before=3926343889, after=3726343889
2015-08-10 13:14:23 update_tally_map(1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb, 2=0x2, +200000000, ttype=0): before=0, after=200000000
2015-08-10 13:14:23 INVESTMENT SEND to Crowdsale Issuer: 1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb
2015-08-10 13:14:23 logicHelper_CrowdsaleParticipation(): granting via crowdsale to user: 0.00003214 2 (TMSC)
2015-08-10 13:14:23 logicHelper_CrowdsaleParticipation(): granting via crowdsale to issuer: 0.00000160 2 (TMSC)
2015-08-10 13:14:23 update_tally_map(1EqTta1Rt8ixAA32DuC29oukbsSWU62qAV, 2147483652=0x80000004, +3214, ttype=0): before=0, after=3214
2015-08-10 13:14:23 update_tally_map(1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb, 2147483652=0x80000004, +160, ttype=0): before=0, after=160
2015-08-10 13:14:23 recordTX(9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d, valid=YES, block= 293017, type= 0, value= 200000000)
```

**Comment by zathras-crypto** [comment]
Nice, good find thanks! :)

I've run through some testing and debugging and can see that the problem is caused by the crowdsale purchase detection in function ```isCrowdsalePurchase()```.  There is duplicate code for main and test ecosystems, and the test ecosystem code contains ```if (sp.issuer == address) continue;``` which should have been a ```!=``` since that line is designed to skip past non matching issuer addresses for optimization.  Instead the only thing we skipped past was the actual matching entry hehe 

I've rewritten that bit to remove the code duplication and fix the bug, should be all sorted in PR https://github.com/OmniLayer/omnicore/pull/238

```
$ src/omnicore-cli omni_gettransaction 9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d
{
    "txid" : "9f8f19ee4dbc6eb23905c6416053a651259a22c88be0e55e61454909d20ce66d",
    "fee" : "0.00020000",
    "sendingaddress" : "1EqTta1Rt8ixAA32DuC29oukbsSWU62qAV",
    "referenceaddress" : "1MCHESTxYkPSLoJ57WBQot7vz3xkNahkcb",
    "ismine" : false,
    "version" : 0,
    "type_int" : 0,
    "type" : "Crowdsale Purchase",
    "propertyid" : 2,
    "divisible" : true,
    "amount" : "2.00000000",
    "purchasedpropertyid" : 2147483652,
    "purchasedpropertyname" : "MCHEST Test Property",
    "purchasedpropertydivisible" : false,
    "purchasedtokens" : "3214",
    "issuertokens" : "160",
    "valid" : true,
    "blockhash" : "0000000000000000c7f1a7fcc5de5bcf90b9ad9774fab78ed925cc60e45b51f3",
    "blocktime" : 1396067389,
    "block" : 293017,
    "confirmations" : 79507
}
```

**Comment by dexX7** [comment]
Pew.. so it was really solely an API/RPC issue. Great job!

**Comment by dexX7** [comment]
Resolved via #238. :)

**Comment by achamely** [comment]
Awesome, thanks guys :)
